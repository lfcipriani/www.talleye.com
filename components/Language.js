import { useRouter } from 'next/router';
import Link from 'next/link';
import i from '../content/i18n';

export default function Language({ lang, alternate, type, wrapper, locale }) {
  const router = useRouter();
  const activeLocale = locale || router.locale;

  function langText(locale, useDefault = locale) {
    return (
      <>
        <img src={`/locales/${locale}.png`} width="16" height="12" aria-hidden="true" />{' '}
        {i('langName', useDefault)[locale]}
        <style jsx>{`
          img {
            display: inline;
          }
        `}</style>
      </>
    );
  }

  function langMessage(text, alternates) {
    return (
      <>
        {text}{' '}
        {alternates.reduce((accu, alt) => {
          return accu === null ? [alt] : [...accu, ', ', alt];
        }, null)}
        .{' '}
      </>
    );
  }

  var text = null;
  var alternates = [];

  if (lang === activeLocale) {
    if (!alternate) {
      return null;
    } else {
      text = <>{i('alsoAvailable')}</>;
    }
  } else {
    alternates.push(langText(lang, activeLocale));
    text = <>{i('onlyAvailable')}</>;
  }
  if (alternate) {
    alternate.forEach((alt) => {
      alternates.push(
        <Link href={`${type ? `/${type}` : ''}/${alt.slug}`} locale={alt.lang}>
          <a>{langText(alt.lang)}</a>
        </Link>
      );
    });
  }

  return (
    <>
      {typeof wrapper === 'function'
        ? wrapper(langMessage(text, alternates))
        : langMessage(text, alternates)}
    </>
  );
}
