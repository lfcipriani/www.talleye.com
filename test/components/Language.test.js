import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Language from '../../components/Language';

describe('locale and content language are equal', () => {
  beforeEach(() => {
    const useRouter = jest.spyOn(require('next/router'), 'useRouter');
    useRouter.mockImplementation(() => ({
      locale: 'en',
      defaultLocale: 'en',
    }));
  });

  it('renders empty without alternate links', () => {
    var { container } = render(<Language lang="en" />);
    expect(container).toBeEmptyDOMElement();
  });

  it('renders links to alternate content', () => {
    const alternate = [{ lang: 'pt-BR', slug: 'slug' }];
    const component = renderer.create(<Language lang="en" alternate={alternate} type="posts" />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders links to 2 alternate contents', () => {
    const alternate = [
      { lang: 'pt-BR', slug: 'slug' },
      { lang: 'de', slug: 'slug' },
    ];
    const component = renderer.create(<Language lang="en" alternate={alternate} type="posts" />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('locale and content language are not equal', () => {
  beforeEach(() => {
    const useRouter = jest.spyOn(require('next/router'), 'useRouter');
    useRouter.mockImplementation(() => ({
      locale: 'pt-BR',
      defaultLocale: 'en',
    }));
  });

  it('render warning without alternate content', () => {
    const component = renderer.create(<Language lang="en" />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('render warning and links to 1 alternate content', () => {
    const alternate = [{ lang: 'pt-BR', slug: 'slug' }];
    const component = renderer.create(<Language lang="en" alternate={alternate} type="posts" />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('render warning and links to 2 alternate contents', () => {
    const alternate = [
      { lang: 'pt-BR', slug: 'slug' },
      { lang: 'de', slug: 'slug' },
    ];
    const component = renderer.create(<Language lang="en" alternate={alternate} type="posts" />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('render with a wrapper', () => {
    const component = renderer.create(
      <Language lang="en" wrapper={(children) => <p data-testid="wrappertest">{children}</p>} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
