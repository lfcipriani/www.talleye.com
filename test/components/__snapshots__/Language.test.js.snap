// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`locale and content language are equal renders links to 2 alternate contents 1`] = `
Array [
  "Also available in",
  " ",
  <span>
    <a
      href="/posts/slug"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <img
        aria-hidden="true"
        className="jsx-3777641437"
        height="12"
        src="/locales/pt-BR.png"
        width="16"
      />
       
      Português
    </a>
  </span>,
  ", ",
  <span>
    <a
      href="/posts/slug"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <img
        aria-hidden="true"
        className="jsx-3777641437"
        height="12"
        src="/locales/de.png"
        width="16"
      />
       
      German
    </a>
  </span>,
  ".",
  " ",
]
`;

exports[`locale and content language are equal renders links to alternate content 1`] = `
Array [
  "Also available in",
  " ",
  <span>
    <a
      href="/posts/slug"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <img
        aria-hidden="true"
        className="jsx-3777641437"
        height="12"
        src="/locales/pt-BR.png"
        width="16"
      />
       
      Português
    </a>
  </span>,
  ".",
  " ",
]
`;

exports[`locale and content language are not equal render warning and links to 1 alternate content 1`] = `
Array [
  "Somente disponível em",
  " ",
  <span>
    <img
      aria-hidden="true"
      className="jsx-3777641437"
      height="12"
      src="/locales/en.png"
      width="16"
    />
     
    Inglês
  </span>,
  ", ",
  <span>
    <a
      href="/posts/slug"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <img
        aria-hidden="true"
        className="jsx-3777641437"
        height="12"
        src="/locales/pt-BR.png"
        width="16"
      />
       
      Português
    </a>
  </span>,
  ".",
  " ",
]
`;

exports[`locale and content language are not equal render warning and links to 2 alternate contents 1`] = `
Array [
  "Somente disponível em",
  " ",
  <span>
    <img
      aria-hidden="true"
      className="jsx-3777641437"
      height="12"
      src="/locales/en.png"
      width="16"
    />
     
    Inglês
  </span>,
  ", ",
  <span>
    <a
      href="/posts/slug"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <img
        aria-hidden="true"
        className="jsx-3777641437"
        height="12"
        src="/locales/pt-BR.png"
        width="16"
      />
       
      Português
    </a>
  </span>,
  ", ",
  <span>
    <a
      href="/posts/slug"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <img
        aria-hidden="true"
        className="jsx-3777641437"
        height="12"
        src="/locales/de.png"
        width="16"
      />
       
      Alemão
    </a>
  </span>,
  ".",
  " ",
]
`;

exports[`locale and content language are not equal render warning without alternate content 1`] = `
Array [
  "Somente disponível em",
  " ",
  <span>
    <img
      aria-hidden="true"
      className="jsx-3777641437"
      height="12"
      src="/locales/en.png"
      width="16"
    />
     
    Inglês
  </span>,
  ".",
  " ",
]
`;

exports[`locale and content language are not equal render with a wrapper 1`] = `
<p
  data-testid="wrappertest"
>
  Somente disponível em
   
  <span>
    <img
      aria-hidden="true"
      className="jsx-3777641437"
      height="12"
      src="/locales/en.png"
      width="16"
    />
     
    Inglês
  </span>
  .
   
</p>
`;
